crates/rune/src/runtime/vm.rs

```diff
@@ -977,23 +976,322 @@ impl Vm {
+    ) -> Result<CallResult<Value>, VmError> {
+        use crate::compile::DynamicFieldMode;
+
+        let index = self.unit.lookup_string(string_slot)?;
+        match target {
+            Value::Object(object) => {
+                let object = object.borrow_ref()?;
+
+                if let Some(value) = object.get(&***index) {
+                    return Ok(CallResult::Ok(value.clone()));
                 }
             }
-        })
+            Value::Struct(typed_object) => {
+                let typed_object = typed_object.borrow_ref()?;
+
+                if let Some(value) = typed_object.get(&***index) {
+                    return Ok(CallResult::Ok(value.clone()));
+                }
+            }
+            Value::Variant(variant) => {
+                let variant = variant.borrow_ref()?;
+
+                if let VariantData::Struct(data) = variant.data() {
+                    if let Some(value) = data.get(&***index) {
+                        return Ok(CallResult::Ok(value.clone()));
+                    }
+                }
+            }
+            target => {
+                return Ok(match target.field_mode()? {
+                    DynamicFieldMode::Never => {
+                        match self.call_field_fn(Protocol::GET, target, index.hash(), ())? {
+                            CallResult::Ok(()) => return Ok(CallResult::Ok(self.stack.pop()?)),
+                            CallResult::Unsupported(target) => CallResult::Unsupported(target),
+                        }
+                    }
+                    #[cfg(feature = "dynamic_fields")]
+                    DynamicFieldMode::First => {
+                        let hash = index.hash();
+                        match self.call_instance_fn(
+                            target.clone(),
+                            Protocol::DYNAMIC_FIELD_GET,
+                            (index.clone(),),
+                        )? {
+                            CallResult::Ok(()) => {
+                                let result = self.stack.pop()?;
+                                if let Value::Option(result) = result {
+                                    let val = result.take()?;
+                                    if let Some(val) = val {
+                                        CallResult::Ok(val)
+                                    } else {
+                                        match self.call_field_fn(Protocol::GET, target, hash, ())? {
+                                            CallResult::Ok(()) => CallResult::Ok(self.stack.pop()?),
+                                            CallResult::Unsupported(target) => {
+                                                CallResult::Unsupported(target)
+                                            }
+                                        }
+                                    }
+                                } else {
+                                    CallResult::Ok(result)
+                                }
```

Comment:
My preferred style in instances like thse would be this to save some of the indentation:

```rust
let result = match result {
    Value::Option(result) => result,
    _ => return Ok(CallResult::Ok(result)),
};
```

Reply:
Normally I agree, was just bit tired during the coding session lol.